cmd/intake-receiver/main.go

```diff
@@ -0,0 +1,331 @@
+	flag.Parse()
+
+	// Create a context that will be cancelled when an interrupt is received.
+	ctx, stop := signal.NotifyContext(context.Background(), os.Interrupt, os.Kill)
+	defer stop()
+
+	if folder != "." {
+		if err := os.MkdirAll(folder, 0755); err != nil {
+			log.Fatalln(err)
+		}
+	}
+
+	var agentFileMap fileMap
+	defer agentFileMap.m.Range(func(_, v interface{}) bool {
+		if f, ok := v.(*syncFile); ok && f != nil {
+			f.Close()
+			log.Println("closed file", f.Name())
+		}
+		return true
+	})
+	rh := requestHandler{
+		agentFileMap: &agentFileMap,
+		basePath:     folder,
+		rootResponse: fmt.Sprintf(`{"publish_ready":true,"version":"%s"}`+"\n", version),
+		bufPool:      sync.Pool{New: func() interface{} { return &bytes.Buffer{} }},
+		bytesBufPool: sync.Pool{New: func() interface{} { return make([]byte, maxScannerBufSize) }},
+	}
+	mux := http.NewServeMux()
+	mux.Handle("/", rh.rootHandler())
+	// TODO(marclop): intake/v2/rum is not supported.
+	for _, p := range []string{"/intake/v2/events", "/intake/v3/rum/events"} {
+		mux.Handle(p, rh.eventHandler())
+	}
+	srv := http.Server{
+		Addr:        host,
+		Handler:     mux,
+		ReadTimeout: 30 * time.Second,
+		BaseContext: func(l net.Listener) context.Context { return ctx },
+	}
+
+	go func() {
+		<-ctx.Done()
+		log.Println("closing http server...")
+
+		shutdownCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		if err := srv.Shutdown(shutdownCtx); err != nil {
+			log.Println(err)
+		}
+	}()
+
+	log.Println("http server listening for requests on", srv.Addr)
+	if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+		log.Println(err)
+	}
+}
+
+type requestHandler struct {
+	agentFileMap *fileMap
+	bufPool      sync.Pool
+	bytesBufPool sync.Pool
+	basePath     string
+	rootResponse string
+}
+
+func (h requestHandler) rootHandler() http.Handler {
+	return logHandler(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		if r.URL.Path == "/" {
+			rw.WriteHeader(200)
+			rw.Write([]byte(h.rootResponse))
+			return
+		}
+		http.Error(
+			rw,
+			fmt.Sprintf(" %s is not implemented", r.URL.Path),
+			http.StatusNotImplemented,
+		)
+	}))
+}
+
+func (h requestHandler) eventHandler() http.Handler {
+	return logHandler(
+		http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			code, err := h.handleRequest(r)
+			if err != nil {
+				log.Println("failed handling request", code, err.Error())
+				http.Error(rw, err.Error(), code)
+				return
+			}
+			rw.WriteHeader(code)
+		}),
+	)
+}
+
+func (h requestHandler) handleRequest(r *http.Request) (int, error) {
+	buf := h.bufPool.Get().(*bytes.Buffer)
+	defer func() {
+		buf.Reset()
+		h.bufPool.Put(buf)
+	}()
+
+	var err error
+	body := r.Body
+	encoding := r.Header.Get("Content-Encoding")
+	switch encoding {
+	case "deflate":
+		body, err = zlib.NewReader(r.Body)
+	case "gzip":
+		body, err = gzip.NewReader(r.Body)
+	}
```

Comment:
default case for unsupported content-encoding? might be pedantic, since we're only supporting the agents whose encoding we control

Reply:
I will add a check for that case, just in case 😄

Reply:
I'm not certain how the RUM agent encodes data; a default option seems reasonable.

Reply:
Just for the sake of the PR review, we assume that the Content-Encoding is "" in the default case in the APM Server intake: https://github.com/elastic/apm-server/blob/2c36ad76b92f92ea89def677d0925cedf5f4f7e3/decoder/req_decoder.go#L56-L83